src/parquet/file/metadata-internal.cc

```diff
@@ -0,0 +1,510 @@
+  }
+
+  void SetStatistics(const ColumnStatistics& val) {
+    format::Statistics stats;
+    stats.null_count = val.null_count;
+    stats.distinct_count = val.distinct_count;
+    stats.max = *val.max;
+    stats.min = *val.min;
+
+    column_chunk_->meta_data.__set_statistics(stats);
+  }
+
+  void Finish(int64_t num_values, int64_t dictionary_page_offset,
+      int64_t index_page_offset, int64_t data_page_offset, int64_t compressed_size,
+      int64_t uncompressed_size) {
+    if (dictionary_page_offset > 0) {
+      column_chunk_->__set_file_offset(dictionary_page_offset + compressed_size);
+    } else {
+      column_chunk_->__set_file_offset(data_page_offset + compressed_size);
+    }
+    column_chunk_->meta_data.__set_dictionary_page_offset(dictionary_page_offset);
+    column_chunk_->meta_data.__set_index_page_offset(index_page_offset);
+    column_chunk_->meta_data.__set_data_page_offset(data_page_offset);
+    column_chunk_->meta_data.__set_total_uncompressed_size(uncompressed_size);
+    column_chunk_->meta_data.__set_total_compressed_size(compressed_size);
+  }
+
+ private:
+  format::ColumnChunk* column_chunk_;
+  std::shared_ptr<WriterProperties> properties_;
+  const ColumnDescriptor* column_;
+};
+
+std::unique_ptr<ColumnMetaDataBuilder> ColumnMetaDataBuilder::GetColumnMetaDataBuilder(
+    std::shared_ptr<WriterProperties> props, const ColumnDescriptor* column,
+    uint8_t* contents) {
+  return std::unique_ptr<ColumnMetaDataBuilder>(
+      new ColumnMetaDataBuilder(props, column, contents));
+}
+
+ColumnMetaDataBuilder::ColumnMetaDataBuilder(std::shared_ptr<WriterProperties> props,
+    const ColumnDescriptor* column, uint8_t* contents)
+    : impl_{std::unique_ptr<ColumnMetaDataBuilderImpl>(
+          new ColumnMetaDataBuilderImpl(props, column, contents))} {}
+
+ColumnMetaDataBuilder::~ColumnMetaDataBuilder() {}
+
+void ColumnMetaDataBuilder::set_file_path(const std::string& path) {
+  impl_->set_file_path(path);
+}
+
+void ColumnMetaDataBuilder::set_encodings(const std::set<Encoding::type>& path) {
+  impl_->set_encodings(path);
+}
+
+void ColumnMetaDataBuilder::Finish(int64_t num_values, int64_t dictionary_page_offset,
+    int64_t index_page_offset, int64_t data_page_offset, int64_t compressed_size,
+    int64_t uncompressed_size) {
+  impl_->Finish(num_values, dictionary_page_offset, index_page_offset, data_page_offset,
+      compressed_size, uncompressed_size);
+}
+
+void ColumnMetaDataBuilder::SetStatistics(const ColumnStatistics& result) {
+  impl_->SetStatistics(result);
+}
+
+class RowGroupMetaDataBuilder::RowGroupMetaDataBuilderImpl {
+ public:
+  explicit RowGroupMetaDataBuilderImpl(std::shared_ptr<WriterProperties> props,
+      const SchemaDescriptor* schema, uint8_t* contents)
+      : properties_(props), schema_(schema), current_column_(0) {
+    row_group_ = reinterpret_cast<format::RowGroup*>(contents);
+    initialize_columns(schema->num_columns());
```

Comment:
initialize the columns since we already know the count from the schema descriptor